packages/circuits/src/utils.ts

```diff
@@ -0,0 +1,256 @@
+import { CircuitConfig } from './CircuitConfig'
+import { SnarkProof } from '@unirep/utils'
+import {
+    EpochKeyControl,
+    Field,
+    ReputationControl,
+    SignupControl,
+    UserStateTransitionControl,
+} from './type'
+
+/**
+ * Format snark proof for verifier smart contract
+ * @param proof The proof of `SnarkProof` type
+ * @returns An one dimensional array of stringified proof data
+ */
+export const formatProofForVerifierContract = (proof: SnarkProof): string[] => {
+    return [
+        proof.pi_a[0],
+        proof.pi_a[1],
+        proof.pi_b[0][1],
+        proof.pi_b[0][0],
+        proof.pi_b[1][1],
+        proof.pi_b[1][0],
+        proof.pi_c[0],
+        proof.pi_c[1],
+    ].map((x) => x.toString())
+}
+
+/**
+ * Format an one dimensional array for `snarkjs` verification
+ * @param proof The string array of the proof
+ * @returns The `SnarkProof` type proof data
+ */
+export const formatProofForSnarkjsVerification = (
+    proof: Field[]
+): SnarkProof => {
+    return {
+        pi_a: [BigInt(proof[0]), BigInt(proof[1]), BigInt('1')],
+        pi_b: [
+            [BigInt(proof[3]), BigInt(proof[2])],
+            [BigInt(proof[5]), BigInt(proof[4])],
+            [BigInt('1'), BigInt('0')],
+        ],
+        pi_c: [BigInt(proof[6]), BigInt(proof[7]), BigInt('1')],
+    }
+}
+
+export const shiftBits = (
+    data: string,
+    shiftBits: bigint,
+    variableBits: bigint
+): bigint => {
+    const _data = BigInt(data)
+    return (_data >> shiftBits) & ((BigInt(1) << variableBits) - BigInt(1))
+}
+
+export const decodeEpochKeyControl = (
+    control: string,
```

Comment:
Could we take a `bigint` here and in the other functions? Changing the types on the `Proof` classes public properties is a breaking change.

Reply:
[1658f01](https://github.com/Unirep/Unirep/pull/547/commits/1658f01f89d47cf25d29314f24ce83aa24a5fddf)